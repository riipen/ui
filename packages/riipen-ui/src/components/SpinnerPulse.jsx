import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import css from "styled-jsx/css";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const SpinnerPulse = ({ classes, color, margin, size }) => {
  const theme = React.useContext(ThemeContext);

  const getStyles = index => {
    return css.resolve`
      @keyframes pulse {
        0% {
          transform: scale(1);
          opacity: 1;
        }
        45% {
          transform: scale(0.1);
          opacity: 0.7;
        }
        80% {
          transform: scale(1);
          opacity: 1;
        }
      }

      .root {
        animation-fill-mode: both;
        animation: pulse 1.5s ${index * 0.12}s infinite
          cubic-bezier(0.2, 0.68, 0.18, 1.08);
        border-radius: 100%;
        display: inline-block;
        height: ${size}px;
        margin: ${margin}px;
        width: ${size}px;
      }

      .primary {
        background-color: ${theme.palette.primary.main};
      }
      .secondary {
        background-color: ${theme.palette.secondary.main};
      }
      .tertiary {
        background-color: ${theme.palette.tertiary.main};
      }
    `;
  };

  return (
    <span className={clsx(classes)}>
      {Array(3)
        .fill(null)
        .map((_, index) => (
          <React.Fragment key={index}>
            <span className={clsx(getStyles(index).className, "root", color)} />
            {getStyles(index).styles}
          </React.Fragment>
        ))}
    </span>
  );
};

SpinnerPulse.displayName = "SpinnerPulse";

SpinnerPulse.propTypes = {
  /**
   * List of additional classes to apply to this component.
   */
  classes: PropTypes.array,

  /**
   * The color of the spinner.
   */
  color: PropTypes.oneOf(["primary", "secondary", "tertiary"]),

  /**
   * The margin in pixels betwene pulse elements.
   */
  margin: PropTypes.number,

  /**
   * The size in pixels the spinner should be.
   */
  size: PropTypes.number
};

SpinnerPulse.defaultProps = {
  classes: [],
  color: "primary",
  margin: 5,
  size: 10
};

export default withClasses(SpinnerPulse);
