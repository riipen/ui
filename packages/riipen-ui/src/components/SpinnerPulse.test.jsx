import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import SpinnerPulse from "./SpinnerPulse";

describe("<SpinnerPulse>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<SpinnerPulse />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<SpinnerPulse />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<SpinnerPulse />);

      expect(wrapper.find("SpinnerPulse").props().classes).toEqual([
        "riipen",
        "riipen-spinnerpulse"
      ]);
      expect(wrapper.find("SpinnerPulse").props().color).toEqual("primary");
      expect(wrapper.find("SpinnerPulse").props().margin).toEqual(5);
      expect(wrapper.find("SpinnerPulse").props().size).toEqual(10);
    });
  });

  describe("classes prop", () => {
    it("applies custom classes", () => {
      const classes = ["classOne"];

      const wrapper = mount(<SpinnerPulse classes={classes} />);

      expect(
        wrapper
          .find("SpinnerPulse")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });
  });

  describe("color prop", () => {
    it("applies valid custom colour class", () => {
      const color = "secondary";

      const wrapper = mount(<SpinnerPulse color={color} />);

      expect(
        wrapper
          .find("SpinnerPulse")
          .childAt(0)
          .childAt(0)
          .hasClass(color)
      ).toEqual(true);
    });

    it("gives an error with invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<SpinnerPulse color="yellow" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("margin prop", () => {
    it("sets valid custom margin", () => {
      const margin = 20;

      const wrapper = mount(<SpinnerPulse margin={margin} />);

      expect(
        wrapper
          .find("JSXStyle")
          .at(0)
          .props().dynamic
      ).toContain(margin);
    });
  });

  describe("size prop", () => {
    it("sets valid custom size", () => {
      const size = 20;

      const wrapper = mount(<SpinnerPulse size={size} />);

      expect(
        wrapper
          .find("JSXStyle")
          .at(0)
          .props().dynamic
      ).toContain(size);
    });
  });
});
