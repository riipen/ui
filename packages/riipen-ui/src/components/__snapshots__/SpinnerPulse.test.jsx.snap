// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpinnerPulse> renders correct snapshot 1`] = `
<withClasses(SpinnerPulse)>
  <SpinnerPulse
    classes={
      Array [
        "riipen",
        "riipen-spinnerpulse",
      ]
    }
    color="primary"
    margin={5}
    size={10}
  >
    <span
      className="riipen riipen-spinnerpulse"
    >
      <span
        className="jsx-4117070321 root primary"
      />
      <JSXStyle
        dynamic={
          Array [
            0,
            10,
            5,
            10,
            "#1f71ae",
            "#f58632",
            "#3caabb",
          ]
        }
        id="4072823428"
      />
      <span
        className="jsx-40703612 root primary"
      />
      <JSXStyle
        dynamic={
          Array [
            0.12,
            10,
            5,
            10,
            "#1f71ae",
            "#f58632",
            "#3caabb",
          ]
        }
        id="4072823428"
      />
      <span
        className="jsx-1373273305 root primary"
      />
      <JSXStyle
        dynamic={
          Array [
            0.24,
            10,
            5,
            10,
            "#1f71ae",
            "#f58632",
            "#3caabb",
          ]
        }
        id="4072823428"
      />
    </span>
  </SpinnerPulse>
</withClasses(SpinnerPulse)>
`;
